superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+Loading of initiating messages can take quite a while, this should not be the default critical path.
+
+Ideally we prefetch the data, and persist it efficiently for later lookups and validation.
+
+We can entirely separate all the data loading from validation:
+
+- **Loading**: Load log-events for each new block, transform into interop-hashes,
+  and append the block of hashes to a append-log.
+- **Validating**: Traverse the append-log, and for each executing hash, check for matching initiating hashes.
+  If at any point invalid, then truncate the log to remove any data that depends on it.
+  *This is quite similar to what the L1-safedb (op-node FP support feature) does*, except we can fetch ahead,
+  and store groups of initiated message hashes rather than L2 blockhashes.
+
+This type of pre-fetched L2 data, in terms of interface is no different from having
+an RPC endpoint that can check a log-event at an block/index and retrieve the cross-unsafe label.
+But since it is pre-fetched, it can serve queries much faster.
+
+### Optimistic verification
+
+Not all L2 chains will synchronously be available and fully verified.
+Optimistically we can pre-fetch all initiating/executing message hashes,
+but verification depends on multiple L2s being available.
+Thus, there may be initiating messages past the "cross-unsafe" progression,
+that we can assume for some level of cross-L2 guarantee, but are not otherwise safe to use yet.
+
+### Messaging Shreds
+
+The idea of streaming partial blocks can be implemented as extension of the above described append-log model;
+the tip of the append-log is maintained by the shreds, and does not have to be fetched anymore.
+This can happen in-memory, with periodic flushes to disk.
+
+### Messaging "Horizon"
+
+When prefetching data, any individual L2 chain must not go too far ahead before other L2s catch up,
+to keep work balanced and verifiable. The "horizon" is the furthest (into future) timestamp that can be seen,
+and can be entirely verified with data up to that point, thanks to the timestamp invariant.
+
+Using this "horizon" property, nodes can implement a form of back-pressure,
+where it is known exactly whether or not all data has been retrieved to verify remaining open cross-L2 dependencies.
+
+### Message safety is not different than block safety
+
+Consider the following block safety levels, and their messaging equivalents:
+
+```
+enum SafetyLevel {
+    INVALID, # conflicting/invalidated data, to be removed from the canonical view
+    UNSAFE,  # not cross-verified anything, yet.
+    CROSS_UNSAFE,  # cross-verified L2
+    SAFE,  # verified L1 also
+    FINAL,  # irreversible
+}
+```
+
+When considering the global log of interop hashes per L2 chain,
+the safety level label just moves along that log, like it does on the chain of blocks.
+
+### Simple interface:
+
+The verifier side is really all just one interface:
+
+```
+type MessageVerifier interface {
+    Check(Identifier, PayloadHash) SafetyLevel
```

Comment:
Couldn't this also be useful for the v0 block builder? It would reduce a lot of diff in op-geth

Reply:
Spoke to proto, yes this will be used for block building